# Excersice 4

Feature selection:

1. You need to load the UCI Arrhythmia dataset
https://archive.ics.uci.edu/ml/datasets/arrhythmia
* the problem is a multi-class classification

2. For comparing the result you need to use kNN (k=3), and compare based on the accuracy with the original dataset (with the all features).

2. Run three different types of feature selection methods (Univariate Statistics, model based, and Iterative Feature Selection).
* The original data set has 279 features.
* You need to run the methods for number_of_features= 20, 50, 100, 150, 200
* All the results should show in the same figure (At the end we have just one figure)

In [1]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import mglearn
from IPython.display import display
from sklearn.exceptions import ConvergenceWarning

from matplotlib import rc
font = {'family' : 'monospace', 'weight' : 'bold', 'size'   : 25}
rc('font', **font) 

plt.rcParams['figure.figsize'] = [20, 10]
plt.rcParams['lines.linewidth'] = 5.0
plt.rcParams['lines.markersize'] = 15.0

import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning) 
warnings.filterwarnings("ignore", category=FutureWarning) 
warnings.filterwarnings("ignore",category=ConvergenceWarning)

In [2]:
#url = https://archive.ics.uci.edu/ml/machine-learning-databases/arrhythmia
columnes = ["Age", "Gender", "Height", "Weight", "QRS_Duration", "PR_Interval",
           "QT_Interval","T_Interval","P_Interval"]
data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/arrhythmia/arrhythmia.data",header=None, na_values = ['?'])


In [3]:
data.rename(columns = {0:'Age', 1:'Gender',2:'Height', 3:'Weight', 4:'QRS_Duration', 5:'PR_Interval',6:'QT_Interval', 7:'T_Interval',8:'P_Interval', 279: 'target' }, inplace = True)


In [4]:
data

Age  Gender  Height  Weight  QRS_Duration  PR_Interval  QT_Interval  \
0     75       0     190      80            91          193          371   
1     56       1     165      64            81          174          401   
2     54       0     172      95           138          163          386   
3     55       0     175      94           100          202          380   
4     75       0     190      80            88          181          360   
..   ...     ...     ...     ...           ...          ...          ...   
447   53       1     160      70            80          199          382   
448   37       0     190      85           100          137          361   
449   36       0     166      68           108          176          365   
450   32       1     155      55            93          106          386   
451   78       1     160      70            79          127          364   

     T_Interval  P_Interval   9  ...  270   271   272  273  274  275  276  \
0           174         121 -16  ...  0.0   9.0  -0.9  0.0  0.0  0.9  2.9   
1           149          39  25  ...  0.0   8.5   0.0  0.0  0.0  0.2  2.1   
2           185         102  96  ...  0.0   9.5  -2.4  0.0  0.0  0.3  3.4   
3           179         143  28  ...  0.0  12.2  -2.2  0.0  0.0  0.4  2.6   
4           177         103 -16  ...  0.0  13.1  -3.6  0.0  0.0 -0.1  3.9   
..          ...         ...  ..  ...  ...   ...   ...  ...  ...  ...  ...   
447         154         117 -37  ...  0.0   4.3  -5.0  0.0  0.0  0.7  0.6   
448         201          73  86  ...  0.0  15.6  -1.6  0.0  0.0  0.4  2.4   
449         194         116 -85  ...  0.0  16.3 -28.6  0.0  0.0  1.5  1.0   
450         218          63  54  ... -0.4  12.0  -0.7  0.0  0.0  0.5  2.4   
451         138          78  28  ...  0.0  10.4  -1.8  0.0  0.0  0.5  1.6   

      277   278  target  
0    23.3  49.4       8  
1    20.4  38.8       6  
2    12.3  49.0      10  
3    34.6  61.6       1  
4    25.4  62.8       7  
..    ...   ...     ...  
447  -4.4  -0.5       1  
448  38.0  62.4      10  
449 -44.2 -33.2       2  
450  25.0  46.6       1  
451  21.3  32.8       1  

[452 rows x 280 columns]

In [5]:
data.isnull().sum().sum()

408

In [6]:
data.describe()

Age      Gender      Height      Weight  QRS_Duration  \
count  452.000000  452.000000  452.000000  452.000000    452.000000   
mean    46.471239    0.550885  166.188053   68.170354     88.920354   
std     16.466631    0.497955   37.170340   16.590803     15.364394   
min      0.000000    0.000000  105.000000    6.000000     55.000000   
25%     36.000000    0.000000  160.000000   59.000000     80.000000   
50%     47.000000    1.000000  164.000000   68.000000     86.000000   
75%     58.000000    1.000000  170.000000   79.000000     94.000000   
max     83.000000    1.000000  780.000000  176.000000    188.000000   

       PR_Interval  QT_Interval  T_Interval  P_Interval           9  ...  \
count   452.000000   452.000000  452.000000  452.000000  452.000000  ...   
mean    155.152655   367.207965  169.949115   90.004425   33.676991  ...   
std      44.842283    33.385421   35.633072   25.826643   45.431434  ...   
min       0.000000   232.000000  108.000000    0.000000 -172.000000  ...   
25%     142.000000   350.000000  148.000000   79.000000    3.750000  ...   
50%     157.000000   367.000000  162.000000   91.000000   40.000000  ...   
75%     175.000000   384.000000  179.000000  102.000000   66.000000  ...   
max     524.000000   509.000000  381.000000  205.000000  169.000000  ...   

              270         271         272         273    274         275  \
count  452.000000  452.000000  452.000000  452.000000  452.0  452.000000   
mean    -0.278982    9.048009   -1.457301    0.003982    0.0    0.514823   
std      0.548876    3.472862    2.002430    0.050118    0.0    0.347531   
min     -4.100000    0.000000  -28.600000    0.000000    0.0   -0.800000   
25%     -0.425000    6.600000   -2.100000    0.000000    0.0    0.400000   
50%      0.000000    8.800000   -1.100000    0.000000    0.0    0.500000   
75%      0.000000   11.200000    0.000000    0.000000    0.0    0.700000   
max      0.000000   23.600000    0.000000    0.800000    0.0    2.400000   

              276         277         278      target  
count  452.000000  452.000000  452.000000  452.000000  
mean     1.222345   19.326106   29.473230    3.880531  
std      1.426052   13.503922   18.493927    4.407097  
min     -6.000000  -44.200000  -38.600000    1.000000  
25%      0.500000   11.450000   17.550000    1.000000  
50%      1.350000   18.100000   27.900000    1.000000  
75%      2.100000   25.825000   41.125000    6.000000  
max      6.000000   88.800000  115.900000   16.000000  

[8 rows x 280 columns]

Using Univariate Statistics

In [7]:
df_data = data.copy()#copy the database

In [8]:
#nan values
df_data.isnull().sum().sum()

408

In [9]:
#list nan cells by columns
mis = df_data.isnull().sum()
mis_df = pd.DataFrame(mis)
mis_df[mis_df[0]!=0]

0
10    8
11   22
12    1
13  376
14    1

Note- column 13 has 376 missing values which out of 452 rows which make more tha 60% of the data, it has no significant for the test, need to be drop.

In [10]:
df_data= df_data.drop(columns=[13])#drop column 13

In [16]:
#fill nan cells of the above columns except 13 by the median of the column
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split

X_df = df_data.iloc[:,:-1] #select all features except target feature
X = np.array(X_df) #convert it to array (Simple Imputer doesn't work with dataframes)
imp = SimpleImputer(missing_values=np.nan, strategy='median') #create imp object to impute median in all missing values 
imp = imp.fit(X) #calculate median values of the features with missing values
X_imp = imp.transform(X) #fill dataset with median values wherever finds missing values
X_imp.shape

(452, 278)

In [17]:
np.isnan(X_imp).sum().sum() #total number of missing values in the dataset

0

In [18]:
y= df_data['target']
X_train, X_test, y_train, y_test = train_test_split(X_imp, y, random_state=0, test_size=.2)
X_train.shape, y_test.shape

((361, 278), (91,))

In [19]:
#print("Test Score with only selected features: {:.3f}".format(clf.score(X_test_selected, y_test)))
#print("Train Score with all features: {:.3f}".format(clf.score(X_train_selected, y_train)))

In [20]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_regression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_selection import chi2

output = pd.DataFrame(index=None, columns=['kn','train_score', 'test_score'])

train_score=[]
test_score=[]

clf = KNeighborsClassifier(n_neighbors=3)
def uni_vari(X_train,X_test, y_train, y_test,kn):
    
    uni_model = SelectKBest(score_func=f_regression, k=kn)
    uni_model.fit(X_train, y_train)

    # transform training set
    X_train_selected = uni_model.transform(X_train)
    X_test_selected = uni_model.transform(X_test)


    clf.fit(X_train, y_train)

    clf.fit(X_train_selected, y_train)

    test_score=clf.score(X_test_selected, y_test)
    train_score=clf.score(X_train_selected, y_train)
    #output=output.append(pd.Series({'kn':n,'train_score':train_score, 'test_score':test_score}),ignore_index=True )
    return test_score, train_score 
        

In [21]:
number_of_features= [20, 50, 100, 150, 200]
for n in number_of_features:
    test_score,train_score=uni_vari(X_train,X_test, y_train, y_test,n)
    output= output.append(pd.Series({'kn':n,'train_score':train_score, 'test_score':test_score}),ignore_index=True )
output

C:\Users\tamra\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:302: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms
C:\Users\tamra\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:302: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms
C:\Users\tamra\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:302: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms
C:\Users\tamra\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:302: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms
C:\Users\tamra\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:302: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms


kn  train_score  test_score
0   20.0     0.717452    0.582418
1   50.0     0.728532    0.604396
2  100.0     0.731302    0.571429
3  150.0     0.739612    0.626374
4  200.0     0.745152    0.615385

Model-Based Feature Selection

In [58]:
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
knn_clf = KNeighborsClassifier()

def model_fun(X_train,X_test, y_train, y_test,kn):
    param_grid = {'n_neighbors':[3],
                 }

    grid_search = GridSearchCV(knn_clf , param_grid,scoring = 'accuracy',return_train_score=True,cv=2)
    grid_search.fit(X_train, y_train)
    SelectFromModel(grid_search,  max_features=kn)
    
    print('test')
    print(grid_search.cv_results_['mean_test_score'])
    print('train')
    print(grid_search.cv_results_['mean_train_score'])

In [59]:
number_of_features= [20, 50, 100, 150, 200]
for n in number_of_features:
    model_fun(X_train,X_test, y_train, y_test,n)

test
[0.60107428]
train
[0.68695519]
test
[0.60107428]
train
[0.68695519]
test
[0.60107428]
train
[0.68695519]
test
[0.60107428]
train
[0.68695519]
test
[0.60107428]
train
[0.68695519]


Iterative Feature Selection

In [38]:
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
output3 = pd.DataFrame(index=None, columns=['kn','train_score', 'test_score'])
def iter_function(X_train,X_test, y_train, y_test,kn):
    select_REF = RFE(RandomForestClassifier(n_estimators=100, random_state=42),
    n_features_to_select=kn)
    select_REF.fit(X_train, y_train)

    X_train_rfe= select_REF.transform(X_train)
    X_test_rfe= select_REF.transform(X_test)

    score = LogisticRegression().fit(X_train_rfe, y_train).score(X_test_rfe, y_test)
    #print("Test score: {:.3f}".format(score))
    score_t = LogisticRegression().fit(X_train_rfe, y_train).score(X_train_rfe, y_train)
    #print("Train score: {:.3f}".format(score_t))
    return score,score_t
   

In [39]:
number_of_features= [20, 50, 100, 150, 200]
for n in number_of_features:
    test_score,train_score=iter_function(X_train,X_test, y_train, y_test,n)
    output3= output3.append(pd.Series({'kn':n,'train_score':train_score, 'test_score':test_score}),ignore_index=True )
#print(output2)
output3

kn  train_score  test_score
0   20.0     0.792244    0.637363
1   50.0     0.853186    0.648352
2  100.0     0.939058    0.593407
3  150.0     0.963989    0.615385
4  200.0     0.983380    0.593407